### PR TITLE
Implement lgamma function

### PR DESCRIPTION
**Pull request by ethanluoyc**
This adds the functions requested in for https://github.com/pytorch/pytorch/issues/678.

### CODE REVIEW

**Comment by colesbury** [comment]
@ethanluoyc, this looks good but is missing the THC (GPU) implementation. Do you have a GPU to test with?

The THC implementation should go in torch/lib/THC/generic/THCTensorMathPointwise.cu

**Comment by ethanluoyc** [comment]
Unfortunately no..Is it possible for you to help me with that? Since it's mentioned that CUDA supports the lgamma function, it should be easy to add lgamma in the GPU code as well.

**Comment by ethanluoyc** [comment]
@colesbury I managed to find where to defined these operations in the THC, added lgamma in the way similar to how other elementwise math operations are defined. I am not sure if there is some registration mechanism needed like in Module.cpp though.

**Comment by ethanluoyc** [comment]
Hmm I dont really understand why this is failing on python 3.5. It appears to be failing for something that I actually didn't modify.

**Review by apaszke — APPROVED** [review]

**Comment by apaszke** [comment]
It seems that you forgot to add `lgamma` to `THCNumerics` struct.

**Comment by ethanluoyc** [comment]
@apaszke Hmm as I don't have a laptop with CUDA I can't test this on my dev environment.
Just made the changes, will this work?

File: torch/lib/THC/THCNumerics.cuh [hunk 1]
```diff
@@ -527,6 +527,7 @@ struct THCNumerics<float> {
   static inline __host__ __device__ bool eq(float a, float b) { return a == b; }
   static inline __host__ __device__ bool ne(float a, float b) { return a != b; }
 
+  static inline __host__ __device__  float lgamma(float a) { return lgamma(a);}
```
Comment (apaszke) [inline torch/lib/THC/THCNumerics.cuh]: it should be `lgammaf` for floats

**Comment by apaszke** [comment]
Yeah this should do it. Just change `lgamma` to `lgammaf` for floats. I'll then run the GPU contbuilds to make sure it all works.

File: torch/lib/THC/THCNumerics.cuh (hunk 1 above)
Comment (ethanluoyc) [inline torch/lib/THC/THCNumerics.cuh]: Do I change both or just the one inside the function body?

**Review by apaszke — APPROVED** [review]

**Review by apaszke — APPROVED** [review]

**Comment by soumith** [comment]
this is now merged into master

**Pull request closed without merging**